### ISSUE TITLE
problems with writing asynchronous http.request and LTN12 sinks/sources

### ISSUE THREAD

**Issue by andreyorst** [lunarmodules/luasocket#421]
Hello!

Sorry if this isn't an appropriate place to discuss this, but I think I found some potential issues with the use of the LTN12 module in your library. I have an issue, so I figured I'll open a ticket and confirm if it is my understanding is lacking, or there's really a problem here.

For context, I'm developing an asynchronous library in pure PUC Lua and wanted to reuse `luasocket.http` to provide an asynchronous HTTP module. I have a working asynchronous TCP module already, and I have read<sup>[1](https://w3.impa.br/~diego/software/luasocket/ltn12.html),[2](http://lua-users.org/wiki/FiltersSourcesAndSinks)</sup> and hopefully understood the LTN12 module, and made some asynchronous sinks and sources. I thought that because the luasocket.http module also uses LTN12 in its `request` function, I'd be able to use an asynchronous sink to delay body consumption until needed.

However, I have noticed that the `request` function can't be used in such a way because it basically ensures that the body is consumed fully before the function returns:

https://github.com/lunarmodules/luasocket/blob/fa69770e52ba869feb8339d49e7c3c536953fbde/src/http.lua#L390-L393

More importantly, I can't access the response headers **until** the body is consumed, which is what I aimed to do. It seems weird, as I may want to discard the body depending on the status code, returned by `request`, but I have to decide upfront if I want to discard it by passing either a sink that discards the output or a sink that collects the output. In the end, the output is always processed, which is good, but I don't think it should happen inside the `request` function.

Here's what I encountered. Initially, I have made the following  sink (*pseudocode* to avoid complex implementation details unrelated to the problem):

```lua
function async_sink (channel)
   return function (chunk)
   if nil ~= chunk then
      channel:put(chunk) -- parks the thread if the channel has no consumers
   else
      channel:close()
   end
   return 1
end
```

All it's necessary to know is that the `put` method parks the current coroutine, which is run by a scheduler if the `channel` is not ready to accept the `chunk`. The `channel` is only ready when there's a pending `take` operation assigned to it, ie. when there's a consumer, possibly in another coroutine (or async thread). Basically, my library implements a form of [CSP](https://en.wikipedia.org/wiki/Communicating_sequential_processes).

Now, by calling `request` like this:

```lua
local response = schedule( -- schedule also returns a channel
  function ()
    local body = new_channel()
    local _, code, headers = http.request(
      {url="http://...",
       sink=async_sink(body)}
    )
    return {status=code,headers=headers,body=body} -- that channel will have this table when ready
  end
)

---8<---
local resp = response:take() -- in some other place of the app on another async thread
if resp.status == 200 then
   content_type = resp.headers["content-type"]
   dispatch[content_type](resp.body:take_all())
else
---8<---
```

The `response` is a channel that, when ready, will contain the table with the status code, headers, and the body channel, which I should be able to consume after I process the response code, and the headers that tell me how to interpret the body. However, the `response:take()` will never succeed.

It will not because the `request` function is calling `ltn12.pump.all(src, snk, step)` in the body, and since we're using an asynchronous sink, which will in turn parks until the `channel` is ready to accept values, that will happen when there's a consumer (and in my case, the consumer appears only after I've parsed the headers) the thread will stay parked, and the `request` will never return the headers. Basically a deadlock. (Yes, my library handles `while` and other loops, that's not the issue here.)

Well, my async runtime can detect simple deadlocks like this, so I figured out what's causing the issue pretty quickly, and tried a different approach by supplying a stepping function that spawns a different asynchronous thread and offloads the pump task to it:

```lua
function async_step(src, snk)
  schedule(function () ltn12.pump.all(src, snk) end) -- offload the pump task
  return nil -- stop the outer pump, releasing the `http.request` function.
end

--8<--
    local _, code, headers = http.request(
      {url="http://...",
       sink=async_sink(body)
       step=async_step}
    )
--8<--
```

This means that when `request` starts consuming the body, its call to `ltn12.pump.all` will exit immediately, spawning a different thread to do the task.

This should work, but then there's a different issue - `request` **closes the connection** when it assumes that the body was consumed by the pump, which hasn't happened yet.
Because of that, the `async_pump` is only able to pump the first chunk, and then it can't get anything from `src`.

<hr>

So, LTN12 are basically streams, except if I understood everything correctly they're not used *as streams* in this context. In my library, channels are also like streams, except they park when the value hasn't arrived and the channel isn't closed. Converting a source or a sink to the channel is as easy as spawning a pump in a separate thread with an appropriate asynchronous sink or asynchronous source as one of its arguments. However, the way sinks and sources are used prevents this kind of interface between these two libraries.

The first question I have - is why the response body has to be a `sink`? Shouldn't it be a `source`, that produces chunks when requested, automatically closing the connection when exhausted, possibly providing a `close` method, that discards the remaining body and closes the connection? This way the user can decide what to do with the response, depending on the status code, it can manually discard it if the response code isn't something they've expected. Passing in a `sink` basically forces current behavior with no ability to workaround it - the only benefit is that the user can pass a `null` sink, which is basically the same as calling `close` on the body if it was a `source`.

I can re-implement the HTTP module as a part of my library, making it work as described, but I thought maybe this was an oversight and it should be instead corrected upstream.

The second question is unrelated to the HTTP module but also prevented me from using LTN12 in an asynchronous context. I've noticed that the `socket.source` with the `"until-closed"` mode behaves strangely when used with clients that had their timeout set to `0`. Again, I have used it with my asynchronous sink, which, when used with other sources, works perfectly, but when I used it with the `socket.source` I couldn't receive any data. Basically, it immediately closes the `source`, as it immediately returns a `nil, "timeout"` error - that's not an `"until-closed"` behavior. Even if there was data, which usually comes after the error, it is discarded in case of the `timeout` error:

https://github.com/lunarmodules/luasocket/blob/fa69770e52ba869feb8339d49e7c3c536953fbde/src/socket.lua#L133-L139

The `partial` only returned when the error message was `"close"`, but `partial` can be returned in case of timeout. I'd expected the implementation to be:

```lua
sourcet["until-closed"] = function(sock)
    local done
    return base.setmetatable({
        getfd = function() return sock:getfd() end,
        dirty = function() return sock:dirty() end
    }, {
        __call = function()
            if done then return nil end
            local chunk, err, partial = sock:receive(socket.BLOCKSIZE)
            if not err then return chunk
            elseif err == "timeout" then
                return partial -- partial is either a non-empty or an empty string but it's there
            elseif err == "closed" then
                sock:close()
                done = 1
                return partial
            else return nil, err end
        end
    })
end
```

It would pollute the sink with `""` empty lines, but they can be filtered out if needed. Are there any particular reasons to close the source at first timeout? I have basically re-implemented this kind of mechanism for my asynchronous TCP module without using sources and sinks, and it works without any issues.

Sorry for this lengthy report - feel free to close it if everything is exactly as intended, hopefully redirecting me to some documents I could read to enhance my understanding in case my assumptions are incorrect. Maybe I'm missing something obvious?

**Comment by andreyorst** [comment]
I've used a *footgun* technique and worked around the first problem by briefly patching the `socket.http` module's `open` function right before I call `request` and resetting it back right after `request` exits, making `open` return the request object with an overridden `close` method.

My idea is that while the body hasn't yet been received this method will close the connection normally, as done when the redirects are handled, for example. But when the asynchronous processing of the body has begun, this method will only pretend to close the connection. And once the body is fully read the connection is closed. The only problem left is an explicit HTTP timeout that I would like to handle via timers instead, but that's already too far as the solution isn't really feasible.

As a proof of concept this seems to work well enough, although it is a major footgun, because I patch the private behavior of the `socket.http` module, which means that if it changes, my library is screwed.

I know that I can pass the `create` function to the `request` function, but this would mean that I have to create a proxy object over the socket, and it messes with the library internals harder than when I override the `open` function. I've tried both approaches, and overriding `open` is clearer, despite being more hacky and dangerous.

Anyway, after a bit more reading of the source code, I see that the `socket.http` is meant to be used synchronously only and the sink interface isn't to provide stream support but to provide control over what to do with the body later, so I guess, my attempts at making it asynchronous are futile.

I'm going to close this ticket, as the overall architecture of this HTTP client isn't meant to be used like that, but feel free to reopen this if you want to discuss it further.
As for the second problem, I guess I can open a separate ticket for that, but granted the synchronous nature, perhaps it's current implementation is fine.
